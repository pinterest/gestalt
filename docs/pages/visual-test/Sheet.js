// @flow strict
import { type Node } from 'react';
import { ColorSchemeProvider, Sheet, Flex, Button, Box, Text } from 'gestalt';

export default function Snapshot(): Node {
  return (
    <ColorSchemeProvider colorScheme="light">
      <Box color="default" display="inlineBlock" width={360} height={360} padding={1}>
        <Sheet
          accessibilityDismissButtonLabel="Test"
          accessibilitySheetLabel="Test"
          footer={
            <Flex justifyContent="end">
              <Button text="Submit" color="red" />
            </Flex>
          }
          heading="Heading"
          onDismiss={() => {}}
          size="sm"
        >
          <Text>Children</Text>
        </Sheet>
      </Box>
    </ColorSchemeProvider>
  );
}
