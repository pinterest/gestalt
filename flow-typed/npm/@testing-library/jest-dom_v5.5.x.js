// flow-typed signature: 5291c1af24f46ab1b593d2626cfbb310
// flow-typed version: 7ad57fc095/@testing-library/jest-dom_v5.5.x/flow_>=v0.104.x

declare module '@testing-library/jest-dom' {
  declare type JestMatcherResult = {
    message?: string | (() => string),
    pass: boolean,
    ...
  };

  declare type Result = JestMatcherResult | Promise<JestMatcherResult>;

  declare module.exports: {|
    /**
     * @deprecated
     */
    toBeInTheDOM(container?: HTMLElement): Result,

    toBeInTheDocument(): Result,
    toBeVisible(): Result,
    toBeEmpty(): Result,
    toBeDisabled(): Result,
    toBeEnabled(): Result,
    toBeInvalid(): Result,
    toBeRequired(): Result,
    toBeValid(): Result,
    toContainElement(element: HTMLElement | null): Result,
    toContainHTML(htmlText: string): Result,
    toHaveAttribute(attr: string, value?: any): Result,
    toHaveClass(...classNames: string[]): Result,
    toHaveFocus(): Result,
    toHaveFormValues(expectedValues: { [name: string]: any, ... }): Result,
    toHaveStyle(css: string | { [name: string]: any, ... }): Result,
    toHaveTextContent(
      text: string | RegExp,
      options?: {| normalizeWhitespace: boolean |}
    ): Result,
    toHaveValue(value?: string | string[] | number): Result,
    toHaveDisplayValue(value: string | string[]): Result,
    toBeChecked(): Result,
  |};
}

declare module '@testing-library/jest-dom/extend-expect' {
  declare module.exports: any;
}
