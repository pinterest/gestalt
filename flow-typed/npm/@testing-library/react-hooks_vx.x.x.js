// flow-typed signature: a58b5bd0827f6c3918d84b9e8a8a52ca
// flow-typed version: <<STUB>>/@testing-library/react-hooks_v^3.4.1/flow_v0.127.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@testing-library/react-hooks'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@testing-library/react-hooks' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@testing-library/react-hooks/dont-cleanup-after-each' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/asyncUtils' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/cleanup' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/flush-microtasks' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/lib/pure' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/pure' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/asyncUtils' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/cleanup' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/flush-microtasks' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src' {
  declare module.exports: any;
}

declare module '@testing-library/react-hooks/src/pure' {
  declare module.exports: any;
}

// Filename aliases
declare module '@testing-library/react-hooks/dont-cleanup-after-each.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/dont-cleanup-after-each'>;
}
declare module '@testing-library/react-hooks/lib/asyncUtils.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/asyncUtils'>;
}
declare module '@testing-library/react-hooks/lib/cleanup.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/cleanup'>;
}
declare module '@testing-library/react-hooks/lib/flush-microtasks.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/flush-microtasks'>;
}
declare module '@testing-library/react-hooks/lib/index' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib'>;
}
declare module '@testing-library/react-hooks/lib/index.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib'>;
}
declare module '@testing-library/react-hooks/lib/pure.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/lib/pure'>;
}
declare module '@testing-library/react-hooks/pure.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/pure'>;
}
declare module '@testing-library/react-hooks/src/asyncUtils.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/asyncUtils'>;
}
declare module '@testing-library/react-hooks/src/cleanup.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/cleanup'>;
}
declare module '@testing-library/react-hooks/src/flush-microtasks.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/flush-microtasks'>;
}
declare module '@testing-library/react-hooks/src/index' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src'>;
}
declare module '@testing-library/react-hooks/src/index.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src'>;
}
declare module '@testing-library/react-hooks/src/pure.js' {
  declare module.exports: $Exports<'@testing-library/react-hooks/src/pure'>;
}
