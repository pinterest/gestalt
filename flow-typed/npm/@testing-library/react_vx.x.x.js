// flow-typed signature: cf9600555f25c805ddd62b5054326bb5
// flow-typed version: <<STUB>>/@testing-library/react_v^11.0.2/flow_v0.127.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@testing-library/react'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@testing-library/react' {
  declare module.exports: any;
}

/**
 * We include stubs for each file inside this npm package in case you need to
 * require those files directly. Feel free to delete any files that aren't
 * needed.
 */
declare module '@testing-library/react/dist/@testing-library/react.cjs' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/@testing-library/react.esm' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/@testing-library/react.pure.cjs' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/@testing-library/react.pure.esm' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/@testing-library/react.pure.umd' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/@testing-library/react.pure.umd.min' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/@testing-library/react.umd' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/@testing-library/react.umd.min' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/act-compat' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/fire-event' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist' {
  declare module.exports: any;
}

declare module '@testing-library/react/dist/pure' {
  declare module.exports: any;
}

declare module '@testing-library/react/dont-cleanup-after-each' {
  declare module.exports: any;
}

declare module '@testing-library/react/pure' {
  declare module.exports: any;
}

// Filename aliases
declare module '@testing-library/react/dist/@testing-library/react.cjs.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.cjs'>;
}
declare module '@testing-library/react/dist/@testing-library/react.esm.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.esm'>;
}
declare module '@testing-library/react/dist/@testing-library/react.pure.cjs.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.pure.cjs'>;
}
declare module '@testing-library/react/dist/@testing-library/react.pure.esm.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.pure.esm'>;
}
declare module '@testing-library/react/dist/@testing-library/react.pure.umd.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.pure.umd'>;
}
declare module '@testing-library/react/dist/@testing-library/react.pure.umd.min.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.pure.umd.min'>;
}
declare module '@testing-library/react/dist/@testing-library/react.umd.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.umd'>;
}
declare module '@testing-library/react/dist/@testing-library/react.umd.min.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/@testing-library/react.umd.min'>;
}
declare module '@testing-library/react/dist/act-compat.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/act-compat'>;
}
declare module '@testing-library/react/dist/fire-event.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/fire-event'>;
}
declare module '@testing-library/react/dist/index' {
  declare module.exports: $Exports<'@testing-library/react/dist'>;
}
declare module '@testing-library/react/dist/index.js' {
  declare module.exports: $Exports<'@testing-library/react/dist'>;
}
declare module '@testing-library/react/dist/pure.js' {
  declare module.exports: $Exports<'@testing-library/react/dist/pure'>;
}
declare module '@testing-library/react/dont-cleanup-after-each.js' {
  declare module.exports: $Exports<'@testing-library/react/dont-cleanup-after-each'>;
}
declare module '@testing-library/react/pure.js' {
  declare module.exports: $Exports<'@testing-library/react/pure'>;
}
