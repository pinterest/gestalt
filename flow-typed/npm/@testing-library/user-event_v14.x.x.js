/*
****** NOTE this is copied from the v13.x.x on flow-typed ******
*
* (https://github.com/flow-typed/flow-typed/blob/e420967f1d88dc042409e9e70ac034ffb968a222/definitions/npm/%40testing-library/user-event_v13.x.x/flow_v0.104.x-/user-event_v13.x.x.js)
* We may run into places where this type isn't correct as the API changed between versions. We should feel free to update these types as needed.
*/

declare module '@testing-library/user-event' {
  // As of Flow 0.134.x WindowProxy is any, which would annihilate all typechecking.
  declare type TargetElement = Element; /* | WindowProxy */

  declare type TypeOpts = {|
    skipClick?: boolean,
    skipAutoClose?: boolean,
    initialSelectionStart?: number,
    initialSelectionEnd?: number,
  |};
  declare type TypeOptsDelay = {|
    ...TypeOpts,
    delay: number,
  |};

  /* Source for keyboard: https://github.com/testing-library/user-event/blob/main/src/keyboard/types.ts */
  declare type KeyboardKey = {|
    code?: string,
    key?: string,
    /** Location on the keyboard for keys with multiple representation
     * STANDARD = 0,
     * LEFT = 1,
     * RIGHT = 2,
     * NUMPAD = 3
     */
    location?: 0 | 1 | 2 | 3,
    keyCode?: number,
    altGr?: boolean,
    shift?: boolean,
  |};
  declare type BaseKeyboardOpts = {|
    document?: Document,
    autoModify?: boolean,
    keyboardMap?: Array<KeyboardKey>,
  |};
  declare type KeyboardState = {|
    pressed: Array<{|
      keyDef: KeyboardKey,
      unpreventedDefault: boolean,
    |}>,
    modifiers: {|
      alt: boolean,
      caps: boolean,
      ctrl: boolean,
      meta: boolean,
      shift: boolean,
    |},
    activeElement: TargetElement | null,
    carryValue?: string,
    carryChar: string,
  |};

  declare type KeyboardOpts = BaseKeyboardOpts;
  declare type KeyboardOptsDelay = {|
    ...BaseKeyboardOpts,
    /** Delay between keystrokes */
    delay: number,
  |};

  declare type TabUserOptions = {|
    shift?: boolean,
    focusTrap?: Document | Element,
  |};

  declare type FilesArgument = File | File[];

  declare type UploadInitArgument = {|
    clickInit?: MouseEvent$MouseEventInit,
    changeInit?: Event,
  |};

  declare type ClickOptions = {|
    skipHover?: boolean,
    clickCount?: number,
  |};

  declare function clear(element: TargetElement): void;

  declare function click(
    element: TargetElement,
    eventInit?: MouseEvent$MouseEventInit,
    options?: ClickOptions
  ): void;

  declare function dblClick(
    element: TargetElement,
    eventInit?: MouseEvent$MouseEventInit,
    options?: ClickOptions
  ): void;

  declare function selectOptions(
    element: TargetElement,
    values: string | string[] | HTMLElement | HTMLElement[],
    eventInit?: MouseEvent$MouseEventInit
  ): void;

  declare function deselectOptions(
    element: TargetElement,
    values: string | string[] | HTMLElement | HTMLElement[],
    eventInit?: MouseEvent$MouseEventInit
  ): void;

  declare function upload(
    element: TargetElement,
    files: FilesArgument,
    init?: UploadInitArgument
  ): void;

  declare function type(
    element: TargetElement,
    text: string,
    userOpts?: TypeOpts
  ): void;
  declare function type(
    element: TargetElement,
    text: string,
    userOpts: TypeOptsDelay
  ): Promise<void>;

  declare function keyboard(
    text: string,
    options?: KeyboardOpts
  ): KeyboardState;
  declare function keyboard(
    text: string,
    options: KeyboardOptsDelay
  ): Promise<KeyboardState>;

  declare function tab(userOpts?: TabUserOptions): void;

  declare function paste(
    element: TargetElement,
    text: string,
    eventInit?: MouseEvent$MouseEventInit,
    pasteOptions?: {|
      initialSelectionStart?: number,
      initialSelectionEnd?: number,
    |}
  ): void;

  declare function hover(
    element: TargetElement,
    init?: MouseEvent$MouseEventInit
  ): void;

  declare function unhover(
    element: TargetElement,
    init?: MouseEvent$MouseEventInit
  ): void;

  declare type UserEvent = {|
    clear: typeof clear,
    click: typeof click,
    dblClick: typeof dblClick,
    deselectOptions: typeof deselectOptions,
    hover: typeof hover,
    paste: typeof paste,
    selectOptions: typeof selectOptions,
    tab: typeof tab,
    type: typeof type,
    keyboard: typeof keyboard,
    unhover: typeof unhover,
    upload: typeof upload,
  |};

  declare export default UserEvent;
}
