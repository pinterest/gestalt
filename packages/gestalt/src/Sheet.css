@keyframes slide-in-ltr {
  from {
    transform: translateX(100%);
  }
  
  to {
    transform: translateX(0%);
  }
}

@keyframes slide-in-rtl {
  from {
    transform: translateX(-100%);
  }

  to {
    transform: translateX(0%);
  }
}

@keyframes slide-out-ltr {
  from {
    transform: translateX(0%);
  }

  to {
    transform: translateX(100%);
  }
}

@keyframes slide-out-rtl {
  from {
    transform: translateX(0%);
  }

  to {
    transform: translateX(-100%);
  }
}

.container {
  composes: fixed borderBox flex justifyEnd left0 top0 itemsCenter from "./Layout.css";
  composes: xsCol12 from "./Column.css";
  height: 100%;
}

.wrapper {
  composes: relative overflowAuto flex from "./Layout.css";
  composes: whiteBgElevated from "./Colors.css";
  height: 100%;
  max-height: 100%;
}

.wrapper:focus {
  outline: none;
}

.wrapperAnimationIn {
  animation: slide-in-ltr 400ms ease-in-out;
}

html[dir="rtl"] .wrapperAnimationIn {
  animation: slide-in-rtl 400ms ease-in-out;
}

.wrapperAnimationOut {
  animation: slide-out-ltr 400ms ease-in-out;
}

html[dir="rtl"] .wrapperAnimationOut {
  animation: slide-out-rtl 400ms ease-in-out;
}

.shadowContainer {
  composes: borderBox relative fit from "./Layout.css";
}

.shadow {
  composes: shadow from "./Borders.css";
}
