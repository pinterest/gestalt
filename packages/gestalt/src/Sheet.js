// @flow strict
import React, {
  forwardRef,
  useCallback,
  useState,
  useEffect,
  useRef,
  type Node,
} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { ESCAPE } from './keyCodes.js';
import { useAnimation } from './AnimationController.js';
import Box from './Box.js';
import Backdrop from './Backdrop.js';
import focusStyles from './Focus.css';
import IconButton from './IconButton.js';
import Heading from './Heading.js';
import Row from './Row.js';
import StopScrollBehavior from './behaviors/StopScrollBehavior.js';
import sheetStyles from './Sheet.css';
import TrapFocusBehavior from './behaviors/TrapFocusBehavior.js';

type Props = {|
  accessibilityDismissButtonLabel: string,
  accessibilitySheetLabel: string,
  children?: Node,
  closeOnOutsideClick?: boolean,
  footer?: Node,
  heading?: string,
  onDismiss: () => void,
  size?: 'sm' | 'md' | 'lg',
|};

const SIZE_WIDTH_MAP = {
  sm: 540,
  md: 720,
  lg: 900,
};

const Header = ({ heading }: {| heading: string |}) => (
  <Box display="flex" justifyContent="start" padding={8}>
    <Heading size="md" accessibilityLevel={1}>
      {heading}
    </Heading>
  </Box>
);

const DismissButton = ({
  accessibilityDismissButtonLabel,
  onClick,
}: {|
  accessibilityDismissButtonLabel: string,
  onClick: () => void,
|}) => (
  <IconButton
    accessibilityLabel={accessibilityDismissButtonLabel}
    bgColor="white"
    icon="cancel"
    iconColor="darkGray"
    onClick={onClick}
  />
);

const SheetWithForwardRef: React$AbstractComponent<
  Props,
  HTMLDivElement
> = forwardRef<Props, HTMLDivElement>(function Sheet(props, sheetRef): Node {
  const {
    accessibilityDismissButtonLabel,
    accessibilitySheetLabel,
    children,
    closeOnOutsideClick = true,
    onDismiss,
    footer,
    heading,
    size = 'sm',
  } = props;

  const [showTopShadow, setShowTopShadow] = useState<boolean>(false);
  const [showBottomShadow, setShowBottomShadow] = useState<boolean>(false);
  const { animationState, onAnimationEnd } = useAnimation();
  const containerRef = useRef<?HTMLDivElement>(null);
  const contentRef = useRef<?HTMLDivElement>(null);

  // Handle onDismiss triggering from ESC keyup event
  useEffect(() => {
    function handleKeyUp(event: {| keyCode: number |}) {
      if (event.keyCode === ESCAPE) {
        onDismiss();
      }
    }

    window.addEventListener('keyup', handleKeyUp);
    return function cleanup() {
      window.removeEventListener('keyup', handleKeyUp);
    };
  }, [onDismiss]);

  // Handle onDismiss triggering from outside click
  const handleOutsideClick = useCallback(() => {
    if (closeOnOutsideClick) {
      onDismiss();
    }
  }, [closeOnOutsideClick, onDismiss]);

  // Handle the shadows on top and bottom of the content area when scrolling
  const updateShadows = useCallback(() => {
    const target = contentRef.current;
    if (!target) {
      return;
    }
    const hasVerticalScrollbar = target.clientHeight < target.scrollHeight;
    setShowTopShadow(hasVerticalScrollbar && target.scrollTop > 0);
    setShowBottomShadow(
      hasVerticalScrollbar &&
        target.offsetHeight + target.scrollTop < target.scrollHeight
    );
  }, []);

  useEffect(() => {
    updateShadows();
    window.addEventListener('resize', updateShadows);
    return () => {
      window.removeEventListener('resize', updateShadows);
    };
  }, [updateShadows]);

  return (
    <StopScrollBehavior>
      <TrapFocusBehavior>
        <div className={sheetStyles.container} ref={containerRef}>
          <Backdrop
            animationState={animationState}
            closeOnOutsideClick={closeOnOutsideClick}
            onClick={handleOutsideClick}
          >
            <div
              aria-label={accessibilitySheetLabel}
              className={classnames(
                sheetStyles.wrapper,
                focusStyles.hideOutline,
                {
                  [sheetStyles.wrapperAnimationIn]: animationState === 'in',
                  [sheetStyles.wrapperAnimationOut]: animationState === 'out',
                }
              )}
              onAnimationEnd={onAnimationEnd}
              ref={sheetRef}
              role="dialog"
              style={{ width: SIZE_WIDTH_MAP[size] }}
              tabIndex={-1}
            >
              <Box
                flex="grow"
                position="relative"
                display="flex"
                direction="column"
                width="100%"
              >
                {heading && (
                  <div
                    className={classnames(sheetStyles.shadowContainer, {
                      [sheetStyles.shadow]: showTopShadow,
                    })}
                  >
                    <Row flex="grow" justifyContent="between">
                      <Box flex="grow">
                        <Header heading={heading} />
                      </Box>
                      <Box flex="none" paddingX={6} paddingY={7}>
                        <DismissButton
                          accessibilityDismissButtonLabel={
                            accessibilityDismissButtonLabel
                          }
                          onClick={onDismiss}
                        />
                      </Box>
                    </Row>
                  </div>
                )}
                {!heading && (
                  <Box
                    display="flex"
                    flex="grow"
                    justifyContent="end"
                    marginBottom={8}
                  >
                    <Box
                      flex="none"
                      paddingX={6}
                      paddingY={7}
                      position="absolute"
                    >
                      <DismissButton
                        accessibilityDismissButtonLabel={
                          accessibilityDismissButtonLabel
                        }
                        onClick={onDismiss}
                      />
                    </Box>
                  </Box>
                )}
                <Box
                  flex="grow"
                  overflow="auto"
                  onScroll={updateShadows}
                  padding={8}
                  ref={contentRef}
                >
                  {children}
                </Box>
                {footer && (
                  <div
                    className={classnames(sheetStyles.shadowContainer, {
                      [sheetStyles.shadow]: showBottomShadow,
                    })}
                  >
                    <Box padding={8}>{footer}</Box>
                  </div>
                )}
              </Box>
            </div>
          </Backdrop>
        </div>
      </TrapFocusBehavior>
    </StopScrollBehavior>
  );
});

// $FlowFixMe Flow(InferError)
SheetWithForwardRef.propTypes = {
  accessibilityDismissButtonLabel: PropTypes.string.isRequired,
  accessibilitySheetLabel: PropTypes.string.isRequired,
  children: PropTypes.node,
  closeOnOutsideClick: PropTypes.bool,
  footer: PropTypes.node,
  heading: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  onDismiss: PropTypes.func,
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
};

SheetWithForwardRef.displayName = 'Sheet';

export default SheetWithForwardRef;
