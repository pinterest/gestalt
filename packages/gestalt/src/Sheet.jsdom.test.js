// @flow strict
import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Sheet from './Sheet.js';
import AnimationControllerModule from './AnimationController.js';

jest.mock('./AnimationController.js');

describe('Sheet', () => {
  const useAnimationMock = jest.spyOn(
    AnimationControllerModule,
    'useAnimation'
  );

  beforeEach(() => {
    useAnimationMock.mockReturnValue({
      animationState: null,
    });
  });

  it('should render all props', () => {
    const sheetRef = React.createRef();
    const { container } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Sheet"
        closeOnOutsideClick
        footer={<footer />}
        heading="Sheet title"
        onDismiss={jest.fn()}
        ref={sheetRef}
        size="sm"
      >
        Sheet content
      </Sheet>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render animation in', () => {
    useAnimationMock.mockReturnValue({
      animationState: 'in',
    });

    const { container } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Sheet"
        onDismiss={jest.fn()}
      >
        Sheet content
      </Sheet>
    );

    expect(container).toMatchSnapshot();
  });

  it('should render animation out', () => {
    useAnimationMock.mockReturnValue({
      animationState: 'out',
    });

    const { container } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Sheet"
        onDismiss={jest.fn()}
      >
        Sheet content
      </Sheet>
    );

    expect(container).toMatchSnapshot();
  });

  it('should attach a ref', () => {
    const sheetRef = React.createRef();

    render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Test Sheet"
        onDismiss={jest.fn()}
        ref={sheetRef}
      >
        Sheet content
      </Sheet>
    );

    expect(sheetRef.current instanceof HTMLDivElement).toEqual(true);
  });

  it('should focus the dismiss button upon render', () => {
    const { container } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Test Sheet"
        onDismiss={jest.fn()}
      >
        Sheet content
      </Sheet>
    );

    expect(container.querySelector('div[role="dialog"]')).toBe(
      document.activeElement
    );
  });

  it('should dismiss from the dismiss button', () => {
    const mockOnDismiss = jest.fn();

    const { getByLabelText } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Test Sheet"
        onDismiss={mockOnDismiss}
      >
        Sheet content
      </Sheet>
    );
    getByLabelText('Dismiss').click();

    expect(mockOnDismiss).toHaveBeenCalledTimes(1);
  });

  it('should dismiss from the ESC key', () => {
    const mockOnDismiss = jest.fn();

    render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Test Sheet"
        onDismiss={mockOnDismiss}
      >
        Sheet content
      </Sheet>
    );
    fireEvent.keyUp(window.document, {
      keyCode: 27,
    });

    expect(mockOnDismiss).toHaveBeenCalledTimes(1);
  });

  it('should dismiss from clicking outside when closeOnOutsideClick is true', () => {
    const mockOnDismiss = jest.fn();

    const { container } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Test Sheet"
        closeOnOutsideClick
        onDismiss={mockOnDismiss}
      >
        Sheet content
      </Sheet>
    );
    const backDrop = container.querySelector('div[role="dialog"]').parentElement
      .firstElementChild;
    fireEvent.click(backDrop);

    expect(mockOnDismiss).toHaveBeenCalledTimes(1);
  });

  it('should not dismiss from the ESC key when closeOnOutsideClick is false', () => {
    const mockOnDismiss = jest.fn();

    const { container } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Test Sheet"
        closeOnOutsideClick={false}
        onDismiss={mockOnDismiss}
      >
        Sheet content
      </Sheet>
    );
    const backDrop = container.querySelector('div[role="dialog"]').parentElement
      .firstElementChild;
    fireEvent.click(backDrop);

    expect(mockOnDismiss).toHaveBeenCalledTimes(0);
  });

  it('should trigger onAnimationEnd', () => {
    const mockOnAnimationEnd = jest.fn();
    useAnimationMock.mockReturnValue({
      animationState: 'post-in',
      onAnimationEnd: mockOnAnimationEnd,
    });

    const { container } = render(
      <Sheet
        accessibilityDismissButtonLabel="Dismiss"
        accessibilitySheetLabel="Test Sheet"
        onDismiss={jest.fn()}
      >
        Sheet content
      </Sheet>
    );
    fireEvent.animationEnd(container.querySelector('div[role="dialog"]'));

    expect(mockOnAnimationEnd).toHaveBeenCalledTimes(1);
  });
});
