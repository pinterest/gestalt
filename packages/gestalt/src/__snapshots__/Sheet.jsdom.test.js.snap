// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sheet should render all props 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="backdrop zoomOut"
      />
      <div
        aria-label="Sheet"
        class="wrapper hideOutline"
        role="dialog"
        style="width: 540px;"
        tabindex="-1"
      >
        <div
          class="box flexGrow relative xsDirectionColumn xsDisplayFlex"
          style="width: 100%;"
        >
          <div
            class="shadowContainer"
          >
            <div
              class="box marginEnd0 marginStart0"
            >
              <div
                class="box flexGrow itemsCenter justifyBetween xsDirectionRow xsDisplayFlex"
              >
                <div
                  class="box"
                >
                  <div
                    class="box flexGrow"
                  >
                    <div
                      class="box paddingX8 paddingY8 xsDisplayFlex"
                    >
                      <h1
                        class="Heading fontSize2 darkGray alignLeft breakWord"
                      >
                        Sheet title
                      </h1>
                    </div>
                  </div>
                </div>
                <div
                  class="box"
                >
                  <div
                    class="box flexNone paddingX6 paddingY7"
                  >
                    <button
                      aria-label="Dismiss"
                      class="button tapTransition enabled"
                      type="button"
                    >
                      <div
                        class="pog white"
                        style="height: 40px; width: 40px;"
                      >
                        <svg
                          aria-hidden="true"
                          aria-label=""
                          class="icon darkGray iconBlock"
                          height="18"
                          role="img"
                          viewBox="0 0 24 24"
                          width="18"
                        >
                          <path
                            d="test-file-stub"
                          />
                        </svg>
                      </div>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="box flexGrow overflowAuto paddingX8 paddingY8"
          >
            Sheet content
          </div>
          <div
            class="shadowContainer"
          >
            <div
              class="box paddingX8 paddingY8"
            >
              <footer />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Sheet should render animation in 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="backdrop backdropAnimationIn zoomOut"
      />
      <div
        aria-label="Sheet"
        class="wrapper hideOutline wrapperAnimationIn"
        role="dialog"
        style="width: 540px;"
        tabindex="-1"
      >
        <div
          class="box flexGrow relative xsDirectionColumn xsDisplayFlex"
          style="width: 100%;"
        >
          <div
            class="box flexGrow justifyEnd marginBottom8 xsDisplayFlex"
          >
            <div
              class="absolute box flexNone paddingX6 paddingY7"
            >
              <button
                aria-label="Dismiss"
                class="button tapTransition enabled"
                type="button"
              >
                <div
                  class="pog white"
                  style="height: 40px; width: 40px;"
                >
                  <svg
                    aria-hidden="true"
                    aria-label=""
                    class="icon darkGray iconBlock"
                    height="18"
                    role="img"
                    viewBox="0 0 24 24"
                    width="18"
                  >
                    <path
                      d="test-file-stub"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
          <div
            class="box flexGrow overflowAuto paddingX8 paddingY8"
          >
            Sheet content
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Sheet should render animation out 1`] = `
<div>
  <div>
    <div
      class="container"
    >
      <div
        class="backdrop backdropAnimationOut zoomOut"
      />
      <div
        aria-label="Sheet"
        class="wrapper hideOutline wrapperAnimationOut"
        role="dialog"
        style="width: 540px;"
        tabindex="-1"
      >
        <div
          class="box flexGrow relative xsDirectionColumn xsDisplayFlex"
          style="width: 100%;"
        >
          <div
            class="box flexGrow justifyEnd marginBottom8 xsDisplayFlex"
          >
            <div
              class="absolute box flexNone paddingX6 paddingY7"
            >
              <button
                aria-label="Dismiss"
                class="button tapTransition enabled"
                type="button"
              >
                <div
                  class="pog white"
                  style="height: 40px; width: 40px;"
                >
                  <svg
                    aria-hidden="true"
                    aria-label=""
                    class="icon darkGray iconBlock"
                    height="18"
                    role="img"
                    viewBox="0 0 24 24"
                    width="18"
                  >
                    <path
                      d="test-file-stub"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
          <div
            class="box flexGrow overflowAuto paddingX8 paddingY8"
          >
            Sheet content
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
